rules/macos/execution_curl_spawned_from_installer_package.toml

```diff
@@ -0,0 +1,71 @@
+[metadata]
+creation_date = "2021/02/23"
+maturity = "production"
+updated_date = "2021/02/23"
+
+[rule]
+author = ["Elastic"]
+description = """
+This rule identifies when the built in macOS Installer program is used to install a package that spawns a shell and runs
+the cURL command.
+"""
+false_positives = [
+    "Custom organization-specific macOS packages that use .pkg files to run cURL could trigger this rule.",
+]
+from = "now-9m"
+index = ["logs-endpoint.events.*"]
+language = "kuery"
+license = "Elastic License"
+name = "Installer Spawning cURL from macOS Package"
+note = "This activity has been observed in the Silver Sparrow campaign. Leverage the Exclusion framework is recommended to tune this to your environment."
+references = ["https://redcanary.com/blog/clipping-silver-sparrows-wings"]
+risk_score = 47
+rule_id = "99239e7d-b0d4-46e3-8609-acafcf99f68c"
+severity = "medium"
+tags = ["Elastic", "Host", "macOS", "Threat Detection", "Execution"]
+timestamp_override = "event.ingested"
+type = "query"
+
+query = '''
+event.category:process and event.type:start and
+ process.parent.executable: "/System/Library/CoreServices/Installer.app/Contents/MacOS/Installer" and
+ process.executable: (/bin/bash or /bin/zsh or /bin/sh) and
+ process.command_line.text: "/usr/bin/curl"
```

Comment:
based on your example, this looks like it would miss due to the lack of wildcards in the `process.command_line.text`. How would you feel about converting this to a sequence rule?

```suggestion
sequence with maxspan=5m
[process where event.type == "start" and 
    process.parent.executable in ("/System/Library/CoreServices/Installer.app/Contents/MacOS/Installer", "/usr/sbin/installer") and 
    process.executable in ("/bin/bash", "/bin/zsh", "/bin/sh")] by process.parent.entity_id
[network where event.type == "start" and process.name == "curl"] by process.entity_id
```

Not sure on the 5m and whether that could be tightened or would need more time. This is a lot less fragile and wouldn't depend on wildcards.

If so, you'd need to update `type` and `language` to `eql`

Reply:
should invert the by process.entity_id, and 5m seems to be too much for an installer to reveal actions (maybe 30s to 1m max):
```
sequence by host.id with maxspan=30s
 [process where event.type == "start" and 
    process.parent.executable in ("/System/Library/CoreServices/Installer.app/Contents/MacOS/Installer", "/usr/sbin/installer") and 
    process.executable in ("/bin/bash", "/bin/zsh", "/bin/sh")] by process.entity_id
 [network where event.type == "start" and process.name == "curl"] by process.parent.entity_id
```

Reply:
fixed, thanks @Samirbous

Reply:
@brokensound77 sorry I made a mistake in my previous comment (network event don't have parent process details), also the curl network event will have a different process.pid/entity_id than the sh/bash process (sh will be parent of curl). maybe we can use a sequence of only processes (drop the network event) Installer -> sh/bash -> curl

```
sequence by with maxspan=30s
 [process where event.type == "start" and 
    process.parent.executable in ("/System/Library/CoreServices/Installer.app/Contents/MacOS/Installer", "/usr/sbin/installer") and process.executable in ("/bin/bash", "/bin/zsh", "/bin/sh")] by process.entity_id
 [process where event.type == "start" and process.name == "curl"] by process.parent.entity_id
```

Reply:
> based on your example, this looks like it would miss due to the lack of wildcards in the `process.command_line.text`. How would you feel about converting this to a sequence rule?
> 
> Not sure on the 5m and whether that could be tightened or would need more time. This is a lot less fragile and wouldn't depend on wildcards.
> 
> If so, you'd need to update `type` and `language` to `eql`


This KQL does detect the activity as is, but I'm not against converting to EQL if that's needed. Since it is an automated process, I didn't observe a span requirement.
```
{
  "_index": ".ds-logs-endpoint.events.process-default-2021.02.16-000007",
  "_type": "_doc",
  "_id": "Ao1ly3cBB5tkWyrvbm0S",
  "_version": 1,
  "_score": null,
  "_source": {
    "agent": {
      "id": "47292c2e-0a1b-c6a7-042d-05abcbdef178",
      "type": "endpoint",
      "version": "7.11.0"
    },
    "process": {
      "Ext": {
        "ancestry": [
          "NDcyOTJjMmUtMGExYi1jNmE3LTA0MmQtMDVhYmNiZGVmMTc4LTI0NTYtMTMyNTg0OTg2ODMuMzQzNTAwMA==",
          "NDcyOTJjMmUtMGExYi1jNmE3LTA0MmQtMDVhYmNiZGVmMTc4LTI0NTQtMTMyNTg0OTg2NzMuMTUyMDIwMDA=",
          "NDcyOTJjMmUtMGExYi1jNmE3LTA0MmQtMDVhYmNiZGVmMTc4LTI0NTQtMTMyNTg0OTg2NzMuOTYwNjAwMA==",
          "NDcyOTJjMmUtMGExYi1jNmE3LTA0MmQtMDVhYmNiZGVmMTc4LTI0NTQtMTMyNTg0OTg2NzMuODg2NjAwMA==",
          "NDcyOTJjMmUtMGExYi1jNmE3LTA0MmQtMDVhYmNiZGVmMTc4LTEtMTMyNTg0ODg2MTQuMA==",
          "NDcyOTJjMmUtMGExYi1jNmE3LTA0MmQtMDVhYmNiZGVmMTc4LTAtMTMyNTg0ODg2MTQuMA=="
        ]
      },
      "args": [
        "/bin/bash",
        "-c",
        "/usr/bin/curl https://specialattributes.s3.amazonaws.com/applications/updater/ver.json > /tmp/version.json"
      ],
      "parent": {
        "args": [],
        "name": "Installer",
        "pid": 2454,
        "args_count": 0,
        "entity_id": "NDcyOTJjMmUtMGExYi1jNmE3LTA0MmQtMDVhYmNiZGVmMTc4LTI0NTYtMTMyNTg0OTg2ODMuMzQzNTAwMA==",
        "executable": "/System/Library/CoreServices/Installer.app/Contents/MacOS/Installer"
      },
      "exit_code": 0,
      "name": "bash",
      "pid": 2456,
      "args_count": 3,
      "entity_id": "NDcyOTJjMmUtMGExYi1jNmE3LTA0MmQtMDVhYmNiZGVmMTc4LTI0NTYtMTMyNTg0OTg2ODMuNDM3ODAwMA==",
      "command_line": "/bin/bash -c /usr/bin/curl https://specialattributes.s3.amazonaws.com/applications/updater/ver.json > /tmp/version.json",
      "executable": "/bin/bash",
      "hash": {
        "sha1": "583297db230f1a68b276690dfb0863f7b0698ccf",
        "sha256": "6de76ab470a16b2a825d223b996d994623473c694c60fccbb71af8691e61c5e0",
        "md5": "95d23ed8b5448779eee9863d2bc5c1ba"
      }
    },
    "message": "Endpoint process event",
    "@timestamp": "2021-02-22T20:18:03.004378Z",
    "ecs": {
      "version": "1.5.0"
    },
    "data_stream": {
      "namespace": "default",
      "type": "logs",
      "dataset": "endpoint.events.process"
    },
    "elastic": {
      "agent": {
        "id": "180845c0-6c9e-11eb-a3e3-b3cc7c78a70f"
      }
    },
    "host": {
      "hostname": "mac-vm.local",
      "os": {
        "Ext": {
          "variant": "macOS"
        },
        "kernel": "Darwin Kernel Version 19.6.0: Tue Jan 12 22:13:05 PST 2021; root:xnu-6153.141.16~1/RELEASE_X86_64",
        "name": "macOS",
        "family": "macos",
        "version": "10.15.7",
        "platform": "macos",
        "full": "macOS 10.15.7"
      },
      "ip": [
        "127.0.0.1",
        "::1",
        "fe80::1",
        "172.16.17.2",
        "fe80::474:6217:9ba7:ff06",
        "fe80::8d9c:1da5:a4aa:f4e9",
        "fe80::8cc7:e7e1:e4e6:b3af"
      ],
      "name": "mac-vm.local",
      "id": "564D3ECC-1EC4-0791-A89A-5CBB687878E7",
      "mac": [
        "00:0c:29:78:78:e7"
      ],
      "architecture": "x86_64"
    },
    "event": {
      "sequence": 17260,
      "ingested": "2021-02-22T20:19:18.158064625Z",
      "created": "2021-02-22T20:18:03.004378Z",
      "kind": "event",
      "module": "endpoint",
      "action": "exec",
      "id": "M1DeQjBIJE7mDob4+++++30v",
      "category": [
        "process"
      ],
      "type": [
        "start"
      ],
      "dataset": "endpoint.events.process"
    },
    "user": {
      "Ext": {
        "real": {
          "name": "variable",
          "id": 501
        }
      },
      "name": "variable",
      "id": 501
    },
    "group": {
      "Ext": {
        "real": {
          "name": "staff",
          "id": 20
        }
      },
      "name": "staff",
      "id": 20
    }
  },
  "fields": {
    "event.ingested": [
      "2021-02-22T20:19:18.158Z"
    ],
    "@timestamp": [
      "2021-02-22T20:18:03.004Z"
    ],
    "event.created": [
      "2021-02-22T20:18:03.004Z"
    ]
  },
  "highlight": {
    "process.command_line.text": [
      "/bin/bash -c /@kibana-highlighted-field@usr@/kibana-highlighted-field@/@kibana-highlighted-field@bin@/kibana-highlighted-field@/@kibana-highlighted-field@curl@/kibana-highlighted-field@ https://specialattributes.s3.amazonaws.com/applications/updater/ver.json > /tmp/version.json"
    ],
    "process.parent.executable": [
      "@kibana-highlighted-field@/System/Library/CoreServices/Installer.app/Contents/MacOS/Installer@/kibana-highlighted-field@"
    ],
    "process.executable": [
      "@kibana-highlighted-field@/bin/bash@/kibana-highlighted-field@"
    ]
  },
  "sort": [
    1614025083004
  ]
}
```

Reply:
> @brokensound77 sorry I made a mistake in my previous comment (network event don't have parent process details), also the curl network event will have a different process.pid/entity_id than the sh/bash process ? (normally sh will be parent of curl). maybe we can use a sequence of only processes (drop the network event) Installer -> sh/bash -> curl
> 
> ```
> sequence by with maxspan=30s
>  [process where event.type == "start" and 
>     process.parent.executable in 
>                     ("/System/Library/CoreServices/Installer.app/Contents/MacOS/Installer", 
>                    "/usr/sbin/installer") and 
>    process.executable in ("/bin/bash", "/bin/zsh", "/bin/sh")
> ] by process.entity_id
>  [process where event.type == "start" and process.name == "curl"] by process.parent.entity_id
> ```
Running this:
```
GET /.ds-logs-endpoint.events.*/_eql/search
{
  "query": """
sequence by with maxspan=30s
 [process where event.type == "start" and 
    process.parent.executable in 
                    ("/System/Library/CoreServices/Installer.app/Contents/MacOS/Installer", 
                   "/usr/sbin/installer") and 
   process.executable in ("/bin/bash", "/bin/zsh", "/bin/sh")
] by process.entity_id
 [process where event.type == "start" and process.name == "curl"] by process.parent.entity_id
  """
}
```
I get this:
```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "parsing_exception",
        "reason" : "line 2:13: mismatched input 'with' expecting {'false', 'not', 'null', 'true', '+', '-', '(', STRING, INTEGER_VALUE, DECIMAL_VALUE, IDENTIFIER, QUOTED_IDENTIFIER}"
      }
    ],
    "type" : "parsing_exception",
    "reason" : "line 2:13: mismatched input 'with' expecting {'false', 'not', 'null', 'true', '+', '-', '(', STRING, INTEGER_VALUE, DECIMAL_VALUE, IDENTIFIER, QUOTED_IDENTIFIER}",
    "caused_by" : {
      "type" : "input_mismatch_exception",
      "reason" : null
    }
  },
  "status" : 400
}
```

Reply:
remove the "by" from the beginning of the sequence:

![image](https://user-images.githubusercontent.com/64742097/109020849-a211c400-76ba-11eb-9b3f-8ed811c37744.png)

Reply:
looks like a typo, try `sequence with maxspan=30s`

Reply:
Thanks for the tweaks Samir, this is what I get for 1am reviews 🙂

Reply:
Got it. That worked, thanks!

Is there a non-JSON way to view this in Discover or would I need to make a Detection Rule? There are 10 hits vs. the KQL's 2. Wanting to check to see if it's just grabbing _more_ info or too much?

Reply:
There's no way to use an EQL search in Discover just yet.